### DISCUSSION TITLE
Authorization Token with Server Components

### DISCUSSION THREAD

**Discussion by wchorski** [keystonejs/keystone#8852] [Questions]
## Stack
- backend: keystone-6
- frontend: Next 13 (with app router)

using Apollo Client's new experimental library [@apollo/experimental-nextjs-app-support](https://github.com/apollographql/apollo-client-nextjs) I'm able to query data with **React Server Components**.

I am trying to figure out how to query data that has any _access filtering_ via Server component.

## Demo

For simplicity, let's say I've set up my **Users** schema with `access.filter.query` set for only _logged in_ users.
I've set up 2 components side by side that accesses **All Users**. 

### 'use client' component that works

`app/users/page.tsx`
```tsx
'use client'
import { gql } from '@apollo/client'
import { useQuery } from '@apollo/experimental-nextjs-app-support/ssr'
import ErrorMessage from '@components/ErrorMessage'
import { User } from '@lib/types'
import { ReactNode } from 'react'
  
type Props = {
  children:ReactNode
}

export default function UsersPage ({ children }:Props) {

  const {data, error, loading} = useQuery(query)
  console.log(data);
  

  if(loading) return <p> loading... </p>
  if(error) return <ErrorMessage error={error} />
  if(!data?.users) return <p> no users found </p>


  return (
    <MAIN_UI_EXAMPLE  />
  )

  
}


const query = gql`
  query Users {
    usersCount
    users {
      id
      name
      nameLast
      email
      role {
        name
      }
    }
  }
`
```

### server component that I can't get working

`app/users_server/page.tsx`
```tsx
import { gql } from '@apollo/client'
import ErrorMessage from '@components/ErrorMessage'
import { fetchUsers } from '@lib/fetchdata/fetchUsers'
import { User } from '@lib/types'
import { ReactNode } from 'react'
  
type Props = {
  children:ReactNode
}

export default async function UsersPage ({ children }:Props) {

  const { users, error } = await fetchUsers()
  

  if(error) return <ErrorMessage error={error} />
  if(!users) return <p> no users found </p>

  console.log({users});
  
  return (
    <MAIN_UI />
  )
}


const query = gql`
  query Users {
    usersCount
    users {
      id
      name
      nameLast
      email
      role {
        name
      }
    }
  }
`
```

`@lib/fetchdata/fetchUsers.ts`
```ts
import { gql } from "@apollo/client"
import { getClient } from "@lib/gqlClient"

export async function fetchUsers(page = 1, perPage = 25 ){

  try {
    const client = getClient()
    const { data } = await client.query({query})

    const { users } = data

    return { users }
    
  } catch (error) {
    console.log('fetch users: ', error);
    return {error}
  }
}

const query = gql`
  query Users {
    usersCount
    users{
      id
      name
      nameLast
      email
      role {
        name
      }
    }
  }
`
```

`@lib/gqlClient.ts`
```ts
import { HttpLink } from "@apollo/client";
import {
  NextSSRInMemoryCache,
  NextSSRApolloClient,
} from "@apollo/experimental-nextjs-app-support/ssr";
import { registerApolloClient } from "@apollo/experimental-nextjs-app-support/rsc";
import { envs } from "@/envs";
import { cookies } from "next/headers";

export const { getClient } = registerApolloClient(() => {

  const cookieStore = cookies()
  const cookieSession = cookieStore.get('keystonejs-session')
  const token = cookieSession?.value
  console.log(`==== Bearer ${token}`);
  

  const client = new NextSSRApolloClient({
    cache: new NextSSRInMemoryCache(),
    link: new HttpLink({
      uri: envs.API_URI,
      fetchOptions: {
        cache: "no-store",
        credentials: 'include',
      },
    }),
    credentials: 'include',
    headers: {
      'Authorization': (token) ? `Bearer ${token}`: "",
      "keep-alive": "true",
      "Apollo-Require-Preflight": "true",
      'Content-Type': 'application/json',
    },
  })

  // console.log(client, null, 2)

  return client
  
})
```
## Final Thoughts

I realize this is more of a Apollo issue, but was hoping to see if there are Keystone specific workarounds here as well. 

I also noticed this example [framework-nextjs-app-directory](https://github.com/keystonejs/keystone/tree/main/examples/framework-nextjs-app-directory). Is this what I should be using instead of trying to retro fit the [framework-nextjs-two-servers](https://github.com/keystonejs/keystone/tree/main/examples/framework-nextjs-two-servers) example?

**Comment by marekryb** [comment]
Hi,

I have not used apollo in a long time nor this package in particular, but nevertheless maybe I can give you some helpful insight.

1. To get actual session token from the cookie, you need to decode it. [This is how it is done for base statelessSessions](https://github.com/keystonejs/keystone/blob/62146aa870e7e921505a437fd8449cbbc841941b/packages/core/src/session/index.ts#L78). If you are using separate backend and frontend you would have to call `Iron.unseal` by yourself. In integrated backend and frontend version, you could probably just call `session.get()` with carefully crafted context (you do not have access to req in page context)
2. In your `@lib/gqlClient.ts` you are mixing static RSC with [dynamic one (SSR)](https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-rendering) which is something that this package is explicitly telling you [not to do](https://github.com/apollographql/apollo-client-nextjs#usage). Since you are using `cookies()` you should only use SSR version. (You can have separate RSC apollo client for pages that do not require authentication).
3. If you are using nextjs anyway, I would suggest following the example with combined backend and frontend. You should just be aware of some limitations: AdminUI is not included (need to run separate instance), DX is a bit worse because you have to restart nextjs when you make some changes on keystone side, some config options are not working because keystone is not in control of the server and probably more that I forgot right now.

EDIT: Thinking again, it might not be easy with separate backend and frontend since the keystone cookie will not be sent to nextjs frontend. Eg.
Nextjs frontend -> localhost:3000
Keystone backend -> localhost:4000
When you visit localhost:3000, your browser will not send (with default settings) cookie for localhost:4000. It will be sent from your client component, since the browser can see request to localhost:4000 and will provide the cookie.
You could make it probably work through having session token stored somewhere - eg. as part of next-auth jwt token. In this case you would send every request to backend with `Authentication: Bearer xxxx` header.
But if you are considering using next-auth then you might consider dropping `@keystone-6/auth` package and do authentication following [this example](https://github.com/keystonejs/keystone/tree/main/examples/custom-session-next-auth).
